### PR TITLE
feat: add proxy support

### LINKED ISSUE #13: feat: add our own http client (with proxy support) and our own http.agent proxy
Related: https://github.com/adobe/aio-cli/issues/224

## Description

1. client: Base it off the `fetch` API. Pay particular attention to the needs of of `swagger-client` custom `userFetch` property: https://github.com/swagger-api/swagger-js/blob/dc84649dd63b7a3a0733b24f828753b070ee2832/docs/usage/http-client.md#custom-fetch This will automatically use our http.agent Proxy below
2. proxy: for https://github.com/adobe/aio-cli/issues/224 - this can be used for example in the `openwhisk` library, where we can't specify our own client, but we can set the proxy/http.agent https://github.com/apache/openwhisk-client-js#constructor-options

## Further Steps

All our `aio` libs, instead of using `node-fetch` or the global `fetch`, should use this implementation instead, and third-party libraries like `swagger-client` and others should use this custom fetch override. Similarly for `openwhisk` and the custom proxy

### PR DESCRIPTION
**Pull request by shazron**
Fixes #13
The module export has changed, thus this is a breaking change and will be a major version bump.

Dependents:
- (use in a Swagger client) https://github.com/adobe/aio-lib-campaign-standard/pull/89
- (use in openwhisk.js) https://github.com/adobe/aio-lib-runtime/pull/69
- (pending) and other libs

## How Has This Been Tested?

1. unit tests
2. server test harness in unit tests (for testing proxy)
3. manual testing via `Dependents` PRs above

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by moritzraho — APPROVED** [review]
lgtm! nice work 🙂

File: src/ProxyFetch.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   * Initialize this class with Proxy auth options
+   *
+   * @param {ProxyAuthOptions} authOptions the auth options to connect with
+   */
+  constructor (authOptions = {}) {
+    logger.debug(`constructor - authOptions: ${JSON.stringify(authOptions)}`)
+    const { proxyUrl, username, password } = authOptions
+
+    const initErrors = []
+    if (!proxyUrl) {
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/ProxyFetch.js:45]: true, this was the result of me removing other required parameters. will fix 😅

**Comment by shazron** [comment]
To facilitate e2e tests in our libraries, we need to extract the test servers so they can be re-used.

Two more things need to be done:
1. Extract the test servers and their tests into their own module
2. We need to create a https-proxy server component to test SSL connections. Currently this is done manually using https://mitmproxy.org/ but we can't automate this, thus we need to create this.

**Comment by moritzraho** [comment]
> fix: add proxy support for exponentialBackoff

Latest changes lgtm

**Comment by shazron** [comment]
> To facilitate e2e tests in our libraries, we need to extract the test servers so they can be re-used.
> 
> Two more things need to be done:
> 
> 1. Extract the test servers and their tests into their own module
> 2. We need to create a https-proxy server component to test SSL connections. Currently this is done manually using https://mitmproxy.org/ but we can't automate this, thus we need to create our own (preferably using existing npm modules). A candidate is https://www.npmjs.com/package/http-mitm-proxy

Both these items are complete. The test servers are internal right now and are in the process of going open source. Once they are open sourced and published, this PR needs to be modified to use it.

**Pull request merged**